File: 1d_diffusion.py
import random
#import sys

#Defined constants
N = 1000       #number of particles
x_cells = 21    #number of cells in x (just a 1D array for now)
end_time = 100  	#number of time steps

steps = [t for t in range(0,end_time)]
cells = [0*n for n in range(0,x_cells)] #1D array of cells, elements intilized to zero
#if x_cells % 2 != 0
#	start_pos1 = int(x_cells/2) + 1
#	cells[start_pos1] = N
#else:
#	start_pos1 = int(x_cells/2)
#	cells[start_pos1]  = N
start_pos1 = 11
cells[start_pos1] = N

print('Dimensions of array: {m}x{n}'.format(m = x_cells, n = 1))
print('The number of time steps: {t}'.format(t = len(steps)))
print('The number of particles: {n}'.format(n = N))
'''
As for the structure to hold the data- can be flexible with this. I've decided 
to put everything into a list. Each element of the list will be a 2-tuple. The
first object will hold the time step, the second object a list of values of the
current distribution of the particles (basically the population of each cell at
that particular time step).

data = [(t1,[0,0,0,10,0,...]),(t2,[0,0,2,5,3,0,...]),...]
'''
data = [(-1,cells)] #need to shift indicies over.

for time in steps:
	'''
	For each time step, compute over all cells.
	Create an empty array for the particles to be added into during randomized
	movement. Once all the particles have been moved, the old distribution is 
	replaced with the new one.
	'''
	cells_current = [0*n for n in range(0,x_cells)]
	for i in range(0,x_cells):
		'''
		The index for the cells is 'i'. Note indexing starts at zero.
		The index ends at (x_cells - 1), (nature of the range function).
		Basically- for each cell- we'll loop over all the particles and randomly
		decide if a particle moves. If it does move, then randomly decide if to 
		the left or to the right. Particles at the first and last cell are 
		subject to boundary conditions, again a random choice decides if the 
		particle stays in place or moves away from the boundary cell.

		Also considering empty cells. Cannot subtract from empty cell. The next
		if suite tests for this. The next iteration of the loop begins if the
		i'th cell is found empty.

		Do not want particles to be handled twice in one time step. I've written
		the program in such a way for each particle movement iteration, a 
		current particle distribution is updated. Basically this preserves
		the 'old' distribution while particles are moved and then replaces
		that distribution.
		'''
		if not cells[i]:
			#go back to the start of the loop if cell is empty
			continue

		for particle in range(0,cells[i]):
			'''
			Processes all the particles in the i'th cell.
			Particles either moved left, right, or stay in current cell. 
			'''
			if random.random() < 0.5:
				#particle will move, next draw decides direction.
				lor = random.random()
				if lor < 0.5 and i != 0 and i != (x_cells - 1): #take this -1 away?):
					#particle will move left, +1 left cell
					cells_current[i-1] += 1
				elif lor > 0.5 and i != 0 and i != (x_cells - 1): #take this -1 away?
					#particle will move right, +1 right cell
					cells_current[i+1] += 1
				
				#the following are the boundary cases
				elif i == 0:
					if lor < 0.5:
						#particle moves to right
						cells_current[i+1] += 1
					else:
						#particle stays in place
						cells_current[i] += 1
				elif i == (x_cells - 1):
					if lor < 0.5:
						#particle moves to left
						cells_current[i-1] += 1
					else:
						#particle stays in place
						cells_current[i] += 1
				else:
					print('error')
			else:
				#particle will not move
				cells_current[i] += 1
	cells = cells_current
	data += [(time,cells)] #10% increase in speed

#writing data to file...
filename = 'particles-{0}_xcells-{1}_time-{2}_startpos-{3}.txt'.format(
			N,x_cells,end_time,start_pos1)	
with open(filename,'w') as f:
	for time_step_data in data:
		f.write(str(time_step_data)+'\n')
print('done')

File: 1d_diffusion.c
/* Diffusion simulation in 1D.*/

#include "stdio.h"
#include "stdlib.h"
#include "math.h"

/* CPP directives to define some constants used in the program.*/
#define N 10000 /*number of particles*/
#define L 1000  /*length of array*/
#define SP 500  /*start position of particles*/

int main(){
	int t, tmax = 5000;
	int i;

	float a = 1.0; /*stepsize*/
	float pl = 0.3, pr = 0.3, ps = 1.0-pl-pr; /*probablities of moving*/
	float rnd;
	float sum_x, sum_x2, avg_x, avg_x2;

	/* The following is an array to hold particle positions. 
	The ith element of that array cooresponds to the current 
	position of the ith particle. This information is never cleared. */
	float x[N] = {0};
	int sxi;

	int x_distribution[L] = {0};
	x_distribution[(int)SP-1] = N; /*setting the start pos. of particles*/

	char *f1 = "/home/paul/Documents/thesis/particle-diffusion/data/\
		1_diffusion_data_P030.txt";
	char *f2 = "/home/paul/Documents/thesis/particle-diffusion/data/\
		1d_diffusion_data_P030-stats.txt";
	FILE *dd, *outstats;
	dd = fopen(f1, "w");
	outstats = fopen(f2, "w");
	
	/*Writing in the initial ditribution...*/
	fprintf(dd, "#First integer is the time step.#\n");
		for(i = 0; i < L; i++){
		fprintf(dd, "%d ", x_distribution[i]);
	}
	fprintf(dd, "\n");


	for (t = 0; t < tmax; t++){
		/*For every time step...*/
		
		for (i = 0; i < L; i++){
			/*Loop over every cell in x_dist. array, set particle count to 0*/
			x_distribution[i] = 0;
		}
		sum_x = 0;
		sum_x2 = 0;
		
		for(i = 0; i < N; i++){
			/*Loop over all particles.*/
			sxi = (int)(x[i] + SP);
			
			/*Check the ith particle position and determine if at boundary.*/
			if(sxi != 0 && sxi != (L-1)){
				/*Particle is not at boundary, may move left, right, or stay.*/
				rnd = (float)rand()/(float)RAND_MAX;
				if(rnd < pl){
					/*The ith particle moves a distance 'a' to the left.*/
					x[i] -= a;
				}
				else if(rnd > pl && rnd < (pl + pr)){
					/*The ith particle moves a distance 'a' to the right.*/
					x[i] += a;
				}
			}
			else if(sxi == 0){
				rnd = (float)rand()/(float)RAND_MAX;
				if(rnd < 0.5){
					/*The ith particle moves a distance 'a' to the right.*/
					x[i] += a;
				}
			}
			else if(sxi == (L-1)){
				rnd = (float)rand()/(float)RAND_MAX;
				if(rnd < 0.5){
					/*The ith particle moves a distance 'a' to the left.*/
					x[i] -= a;
				}
			}
			/**/
			sum_x += x[i];
			sum_x2 += x[i]*x[i];

			sxi = (int)(x[i] + SP); /*Shift the position of the ith particle*/
			x_distribution[sxi]++; 
		}
		
		fprintf(dd, "%d ",t); /*first column is time*/
		for(i = 0; i < L; i++){
			fprintf(dd, "%d ", x_distribution[i]);
		}
		fprintf(dd,"\n");
		
		/*Writing some stats to file...*/
		avg_x = sum_x/(float)N;
		avg_x2 = sum_x2/(float)N;
		fprintf(outstats, "%f %f %f\n", avg_x, avg_x2, avg_x2-avg_x*avg_x);
	}
}